### PR TITLE
18 update changes from crypto3 zk

### PR DESCRIPTION
**Pull request by vo-nil**
This PR contains synchronization of current work at crypto3-zk with actor-zk
Additions include algorithmic performance improvements as well as addition of new mask parameter.

### CODE REVIEW

File: test/systems/plonk/placeholder/placeholder.cpp [hunk 1]
```diff
@@ -615,13 +622,17 @@ ACTOR_THREAD_TEST_CASE(permutation_argument_test) {
     for (int i = 0; i < argument_size; i++) {
         BOOST_CHECK(prover_res.F_dfs[i].evaluate(y) == verifier_res[i]);
         for (std::size_t j = 0; j < desc.rows_amount; j++) {
-            BOOST_CHECK(
-                prover_res.F_dfs[i].evaluate(preprocessed_public_data.common_data.basic_domain->get_domain_element(j)) == field_type::value_type::zero());
+            if(prover_res.F_dfs[i].evaluate(preprocessed_public_data.common_data.basic_domain->get_domain_element(j)) != 0){
```
Comment (Zerg1996) [inline test/systems/plonk/placeholder/placeholder.cpp:625]: It's something interesting. 
We always check inside prover values with `field_type::value_type::zero()`
Such case means - our equation not always correct?

**Review by Zerg1996 — COMMENTED** [review]
I have minor code style notes.

File: include/nil/actor/zk/commitments/polynomial/kzg.hpp [hunk 2]
```diff
@@ -714,11 +714,17 @@ namespace nil {
                         return result;
                     }
 
-                    void setup(transcript_type& transcript){
+                    using preprocessed_data_type = bool;
+                    preprocessed_data_type preprocess(transcript_type& transcript) const{
```
[the remark this answers is not in the document]
Comment (ETatuzova) [reply include/nil/actor/zk/commitments/polynomial/kzg.hpp:718]: It's dummy functions just for buildability.

File: include/nil/actor/zk/commitments/polynomial/lpc.hpp [hunk 3]
```diff
@@ -63,35 +65,52 @@ namespace nil {
...
+                            result[index] = {};
+                            for (const auto& poly: this->_polys.at(index)){
+                                result[index].push_back(poly.evaluate(etha));
+                            }
+                        }
+                        return result;
                     }
 
-                    void setup(transcript_type& transcript){
+                    void setup(transcript_type& transcript, const preprocessed_data_type &preprocessed_data) {
```
[the remark this answers is not in the document]
Comment (ETatuzova) [reply include/nil/actor/zk/commitments/polynomial/lpc.hpp:96]: Yes, transcript.challege() changes transcript's state

**Comment by vo-nil** [comment]
This PR contains synchronization of current work at crypto3-zk with actor-zk
Additions include algorithmic performance improvements as well as addition of new mask parameter.

File: include/nil/actor/zk/snark/arithmetization/plonk/detail/lookup_table_definition.hpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                        );
+                        std::vector<plonk_column<FieldType>> selector_columns;
+
+                        std::size_t start_row = 1;
+                        std::size_t table_index = 0;
+                        std::vector<plonk_lookup_table<FieldType>> bp_lookup_tables(lookup_table_ids.size());
+                        for( const auto&[k, table]:lookup_tables ){
+                            // Place table into constant_columns.
+                            for( std::size_t i = 0; i < table->get_table().size(); i++ ){
+                                if(constant_columns[i].size() < start_row + table->get_table()[i].size()){
```
[the remark this answers is not in the document]
Comment (vo-nil) [reply include/nil/actor/zk/snark/arithmetization/plonk/detail/lookup_table_definition.hpp]: updated in  ab70e489

File: include/nil/actor/zk/commitments/polynomial/lpc.hpp [hunk 5]
```diff
@@ -63,35 +65,52 @@ namespace nil {
                     using poly_type = PolynomialType;
                     using lpc = LPCScheme;
                     using eval_storage_type = typename LPCScheme::eval_storage_type;
+                    using preprocessed_data_type = std::map<std::size_t, std::vector<typename field_type::value_type>>;
```
[the remark this answers is not in the document]
Comment (vo-nil) [reply include/nil/actor/zk/commitments/polynomial/lpc.hpp]: updated in 78fb71c6

File: include/nil/actor/zk/math/non_linear_combination.hpp [hunk 6]
```diff
@@ -129,7 +136,10 @@ namespace nil {
                     }
                     this->terms.clear();
                     for (const auto& it: unique_terms) {
-                        this->terms.emplace_back(it.first.get_vars(), it.second);
+                        if (it.second != assignment_type::zero())
+                        {
```
[the remark this answers is not in the document]
Comment (vo-nil) [reply include/nil/actor/zk/math/non_linear_combination.hpp]: updated in 2c7efa01

File: test/systems/plonk/placeholder/placeholder.cpp (hunk 1 above)
Comment (vo-nil) [reply test/systems/plonk/placeholder/placeholder.cpp]: It seems that some debug checks creeped in.
fixed in c1c33e86

**Review by Zerg1996 — APPROVED** [review]

**Pull request merged**